### PR TITLE
Add savepoint support to atomic distributed transaction

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description

This PR adds support for savepoints usage in atomic distributed transaction commit by storing the savepoints in the transaction query log and then reapplying them on a recovery.

## Related Issue(s)

- #16245 

## Checklist

-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/test/endtoend/transaction/twopc/twopc_test.go [hunk 1]
```diff
@@ -1023,6 +1031,304 @@ func TestReadingUnresolvedTransactions(t *testing.T) {
 	}
 }
 
+// TestDTSavepointWithVanilaMySQL tests distributed transaction should work with savepoint as with vanila MySQL
```
Comment (systay) [inline go/test/endtoend/transaction/twopc/twopc_test.go:1034]:
```suggestion
// TestDTSavepointWithVanillaMySQL ensures that distributed transactions 
// work with savepoints, consistent with vanilla MySQL behavior.
```

File: go/test/endtoend/transaction/twopc/twopc_test.go [hunk 2]
```diff
@@ -1023,6 +1031,304 @@ func TestReadingUnresolvedTransactions(t *testing.T) {
...
+func TestDTSavepointWithVanilaMySQL(t *testing.T) {
+	conn, closer := start(t)
+	defer closer()
+
+	// internal savepoint
+	utils.Exec(t, conn, "begin")
+	utils.Exec(t, conn, "insert into twopc_user(id, name) values(7,'foo'), (8,'bar')")
+	utils.Exec(t, conn, "commit")
+	utils.Exec(t, conn, "select * from twopc_user order by id")
```
Comment (systay) [inline go/test/endtoend/transaction/twopc/twopc_test.go:1043]: shouldn't we be doing assertions here?

File: go/vt/vttablet/tabletserver/tx/api.go [hunk 3]
```diff
@@ -60,13 +59,13 @@ type (
 
 		Stats *servenv.TimingsWrapper
 	}
+	Query struct {
```
Comment (systay) [inline go/vt/vttablet/tabletserver/tx/api.go:62]: maybe add some comments? the type is not self evident, imho

File: go/vt/vttablet/tabletserver/tx/api.go [hunk 4]
```diff
@@ -181,3 +204,9 @@ func (p *Properties) String(sanitize bool, parser *sqlparser.Parser) string {
 		printQueries(),
 	)
 }
+
+func (p *Properties) GetQueries() []Query {
```
Comment (systay) [inline go/vt/vttablet/tabletserver/tx/api.go:208]: the other methods are safe against `p` being nil, but this method is not. is that intentional?

**Review by systay — APPROVED** [review]
apart from some nitpicks, LGTM

**Review by GuptaManan100 — APPROVED** [review]
This is great!

**Pull request merged**